File: HashMap/Subarrays with equal 1's and 0's/Solution.cpp
class Solution{
public:

    long long int countSubarrWithEqualZeroAndOne(int arr[], int n){
        int ans = 0;

        unordered_map<int, int> map;
        map[0] = 1; //add the sum 0 in the map

        int sum = 0;

        for(int i=0; i<n; i++){
            //if value at any index is 0, add -1 to the sum or add 1
            if(arr[i] == 0){
                sum += -1;
            } else{
                sum += 1;
            }

            //if the sum already exists in the map, add its frequency to the ans
            if(map.find(sum) != map.end()){
                ans += map[sum];
            }

            //update the frequency of the sum
            map[sum]++;
        }

        return ans;
    }
};

File: HashMap/Subarrays with equal 1's and 0's/Solution.java
import java.util.HashMap;

class solution{
    public int countSubarrWithEqualZeroAndOne(int arr[], int n){
        int ans = 0;

        HashMap<Integer, Integer> map = new HashMap<>();

        map.put(0, 1); //initaialize the sum 0 with freq 1
        int sum = 0;

        //check for every value in the arr
        for(int val : arr){
            //if value is 0 add -1 to the sum
            if(val == 0){
                sum += -1;
            } else{
                sum += 1; //otherwisw, add +1
            }

            //if that sum already exist in the map (repeated)
            if(map.containsKey(sum)){
                //add the frequency of the sum to the ans
                ans += map.get(sum);
                //update the frequency
                map.put(sum, map.get(sum) +1);
            } else{
                map.put(sum, 1); //otherwise put the freq = 1
            }
        }

        return ans;
    }

};